push/src/main/java/org/aerogear/mobile/push/PushService.java

```diff
@@ -0,0 +1,313 @@
+                    }
+
+                    firebaseMessaging.unsubscribeFromTopic(unifiedPushCredentials.getVariant());
+
+                    callback.onSuccess();
+                    break;
+                default:
+                    callback.onError(new HttpException());
+                    break;
+            }
+        });
+
+    }
+
+    private String getHashedAuth(String username, char[] password) {
+        StringBuilder headerValueBuilder = new StringBuilder("Basic").append(" ");
+        String unhashedCredentials = username + ":" + String.valueOf(password);
+        String hashedCrentials =
+                        Base64.encodeToString(unhashedCredentials.getBytes(), Base64.NO_WRAP);
+        return headerValueBuilder.append(hashedCrentials).toString();
+    }
+
+
+    /**
+     * When a push message is received, all main thread handlers will be notified on the main(UI)
+     * thread. This is very convenient for Activities and Fragments.
+     *
+     * @param handler a handler to added to the list of handlers to be notified.
+     */
+    public static void registerMainThreadHandler(MessageHandler handler) {
+        MAIN_THREAD_HANDLERS.add(handler);
+    }
+
+    /**
+     * When a push message is received, all background thread handlers will be notified on a non UI
+     * thread. This should be used by classes which need to update internal state or preform some
+     * action which doesn't change the UI.
+     *
+     * @param handler a handler to added to the list of handlers to be notified.
+     */
+    public static void registerBackgroundThreadHandler(MessageHandler handler) {
+        BACKGROUND_THREAD_HANDLERS.add(handler);
+    }
+
+    /**
+     * This will remove the given handler from the collection of main thread handlers. This MUST be
+     * called when a Fragment or activity is backgrounded via onPause.
+     *
+     * @param handler a handler to be removed to the list of handlers
+     */
+    public static void unregisterMainThreadHandler(MessageHandler handler) {
+        MAIN_THREAD_HANDLERS.remove(handler);
+    }
+
+    /**
+     * This will remove the given handler from the collection of background thread handlers.
+     *
+     * @param handler a handler to be removed to the list of handlers
+     */
+    public static void unregisterBackgroundThreadHandler(MessageHandler handler) {
+        BACKGROUND_THREAD_HANDLERS.remove(handler);
+    }
+
+    /**
+     * This will deliver an message to all registered handlers.
+     *
+     * @param message the message to pass
+     */
+    public static void notifyHandlers(final Context context, final Map<String, String> message) {
+
+        if (BACKGROUND_THREAD_HANDLERS.isEmpty() && MAIN_THREAD_HANDLERS.isEmpty()
+                        && defaultHandler != null) {
+            new Thread(new Runnable() {
+                public void run() {
+                    defaultHandler.onMessage(context, message);
```

Comment:
if the caller will change the content of the map during this thread execution, we could get a threading issue. Consider copying to passed in map before working with it.

Reply:
Map should not be changed

Reply:
Yes, but you don't have control over that, because the caller has reference to the Map.
To avoid issues, it is a good practice to copy its content so that you don't have to worry about what the caller will do with the Map.

Reply:
Sorry I got your best practice point here. My point is this care I really don't care if anyone changes it because the Services responsibility is only to pass it along and go home. I mean the lib is not going to use it anywhere anymore. Anyway, I'm not against to do a defensive copy of this.

Reply:
Yes, but what would happen in the for loop we have just a few lines below? We could end up passing e different `message` to each `handler` (line 262)

Reply:
Ok, I have added `Collections.unmodifiableMap(message)` so!